sql/catalog: consolidating index resolving logic into catalog/resolver package.

chengxiong-ruan opened the pull request:
This PR contains 3 commits:
(1) extract `SchemaResolver` interface implementation of planner into a separate struct but still within the sql package. and embed it into planner.
(2) consolidate index resolving logic in `opt/cat` and `sql` packages.
(3) add `MayResolveIndex` method to `scbuild.CatalogReader` interface.

chengxiong-ruan committed:
psql: extract schema resolver implemtation of planner into separate struct

Extract schema resolver interface implementation of planner into
a separate struct. This would make the index resolution refactoring
easier for new schema changer.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/80769)
<!-- Reviewable:end -->

chengxiong-ruan committed:
resolver: consolidate index resolution logic

Previously, we have very similar logics to resolve Index with
a TableIndexName in `opt` and in `sql` packages. This commit
consolidate them into the resolver package.

Release note: None.

chengxiong-ruan committed:
sql/schemachanger: add index resolving interface method

This commit simply utilizes the consolidated index resolving
function to add a `MayResolveIndex` interface to the CatalogReader
interface, so that it can be used to resolve index elements
for scenarios like `DROP INDEX`, `ALTER INDEX` or `COMMENT ON INDEX`.

Release note: None.

ajwerner commented on pkg/sql/schema_resolver.go:
```diff
@@ -0,0 +1,337 @@
+
+var _ resolver.SchemaResolver = &schemaResolver{}
+
+// This is might be a bad name.
```
I think it's a fine name, but it's a bad comment

ajwerner commented on pkg/sql/schema_resolver.go:
```diff
@@ -0,0 +1,337 @@
+	// skipDescriptorCache, when true, instructs all code that accesses table/view
+	// descriptors to force reading the descriptors within the transaction. This
+	// is necessary to read descriptors from the store for:
+	// 1. Descriptors that are part of a schema change but are not modified by the
+	// schema change. (reading a table in CREATE VIEW)
+	// 2. Disable the use of the table cache in tests.
```
uber-nit: indent this list by a couple of space?

ajwerner commented on pkg/sql/schema_resolver.go:
```diff
@@ -0,0 +1,337 @@
+var _ resolver.SchemaResolver = &schemaResolver{}
+
+// This is might be a bad name.
+type schemaResolver struct {
```
what's holding us back from moving this out to a separate package? Is it just that you've got those planner methods which set these flags? Seems straight-forward to just run some setter methods.

One approach to avoid the allocation is to put an init method on the thing or a `MakeSchemaResolver` or something which returns the struct value directly.

ajwerner reviewed (commented):
I need to give the second commit a closer read

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80769)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[`pkg/sql/schema_resolver.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80769#-N0nEueeRwJiFMVAhgxb-r1-37:-N1AB9j83cnBBnUWVbph:b44t7x5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251b98c8123069ea35429200d75089c794cc8d6/pkg/sql/schema_resolver.go#L37)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think it's a fine name, but it's a bad comment
</blockquote></details>

OMG, I don't even remember I added this bad comment... thanks for catching it. Fixed.
___
*[`pkg/sql/schema_resolver.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80769#-N0nEueeRwJiFMVAhgxb-r1-38:-N1ABFKc7Snzg9g50GwH:b9dccue) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251b98c8123069ea35429200d75089c794cc8d6/pkg/sql/schema_resolver.go#L38)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

what's holding us back from moving this out to a separate package? Is it just that you've got those planner methods which set these flags? Seems straight-forward to just run some setter methods.

One approach to avoid the allocation is to put an init method on the thing or a `MakeSchemaResolver` or something which returns the struct value directly. 
</blockquote></details>

yeah, but not just the flags, but also the txn and sessionDataStack which must be reset when planner is reset..
But I agree that we may just expose these through public methods and use it as explicit type instead of a SchemaResolver interface in planner.
I'll do it with as a follow up in a separate pr if you don't mind.
___
*[`pkg/sql/schema_resolver.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80769#-N0nEueeRwJiFMVAhgxb-r1-49:-N1ABLDb1hUFjmZ4A0-r:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251b98c8123069ea35429200d75089c794cc8d6/pkg/sql/schema_resolver.go#L49)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

uber-nit: indent this list by a couple of space?
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/schema_resolver.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80769#-N0nEueeRwJiFMVAhgxb-r1-38:-N1ABVlV1ESqaUOZpwT6:b-hpbp3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251b98c8123069ea35429200d75089c794cc8d6/pkg/sql/schema_resolver.go#L38)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

yeah, but not just the flags, but also the txn and sessionDataStack which must be reset when planner is reset..
But I agree that we may just expose these through public methods and use it as explicit type instead of a SchemaResolver interface in planner.
I'll do it with as a follow up in a separate pr if you don't mind.
</blockquote></details>

oh, I forgot the `canResolveDescUnderSchema` thing need to be extracted to somewhere, either exposed as public method from the new struct or some helper function somewhere.


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/catalog/resolver/resolver.go:
```diff
@@ -572,3 +540,295 @@ func ResolveObjectNamePrefix(
+// contain indexes with same name.
+// (3) table name is present from input, but the resolve object is not a table
+// ore materialized view.
+// (4) table name is not present, the given schema or schemas on search path can
+// not be found.
```
this last one is interesting, is it valid? What use case is it supporting?

ajwerner commented on pkg/sql/catalog/resolver/resolver.go:
```diff
@@ -572,3 +540,295 @@ func ResolveObjectNamePrefix(
+	return resolvedPrefix, nil
+}
+
+// findTableContainsIndex tries to find and index in a schema by iterating
+// through all tables.
+// Error is returned if Index name is ambiguous between tables. No error is
+// returned if no table is found.
+func findTableContainsIndex(
```
nit: how about `findTableContainingIndex`?

ajwerner reviewed (approved):
Nice job wading through a real mess. Eventually i'd love to see us further clean up the `catalog/resolver` code to be a more testable layer than it presently is, but one step at a time.

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80769)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[`pkg/sql/catalog/resolver/resolver.go` line 559 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80769#-N0xSvRBzOm8TSnz4dTU-r6-559:-N1EcrZe7etaD9_C688r:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/f42d318e900203baa8f4dfa5634db4cd22df19f1/pkg/sql/catalog/resolver/resolver.go#L559)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: ore
</blockquote></details>

fixed
___
*[`pkg/sql/catalog/resolver/resolver.go` line 561 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80769#-N0xSvRBzOm8TSnz4dTU-r6-561:-N1EcsDz3WK-MoQFZPst:b4lqjvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f42d318e900203baa8f4dfa5634db4cd22df19f1/pkg/sql/catalog/resolver/resolver.go#L561)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this last one is interesting, is it valid? What use case is it supporting?
</blockquote></details>

this happens for example when you set `search_path` to some non-exist schema, which should be rare, but we do have test for this.
___
*[`pkg/sql/catalog/resolver/resolver.go` line 783 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80769#-N0xSvRBzOm8TSnz4dTU-r6-783:-N1Ed7ym3xq0a8A2Ki-p:b48erpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f42d318e900203baa8f4dfa5634db4cd22df19f1/pkg/sql/catalog/resolver/resolver.go#L783)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: how about `findTableContainingIndex`?
</blockquote></details>

good choice.
___
*[`pkg/sql/opt/cat/catalog.go` line 131 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80769#-N0xSvRBzOm8TSnz4dTX-r6-131:-N1EeUj1-p4IVTEbpwLy:bui5r3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f42d318e900203baa8f4dfa5634db4cd22df19f1/pkg/sql/opt/cat/catalog.go#L131)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

give this commentary?
</blockquote></details>

added, thx!


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
test added and rebased

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80769#-:-N1FjyUe0EVLz7jNpjaX:b-d0fhsi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
TFTR!
bors r+
